java.lang.IllegalStateException: Toolkit not initialized

I'm just capturing this issue for discussion as in general the infamous "Toolkit not initialized" shows up many different ways, and I see there still is no javafx.application.Platform.isRunning() type method to check if the platform has initialized. 

In this particular case I have a `@Scheduled` activity that I can simply add an initial delay to, but in general, how CDI components interact with javafx during startup needs hacks to wait for initialization.

Maybe there needs to be a quarkus-fx runnable event queue that can take care of delaying submitting the tasks until the application is fully running.

```
2024-02-05 14:04:45,011 ERROR [io.qua.sch.com.run.StatusEmitterInvoker] (executor-thread-1) Error occurred while executing task for trigger IntervalTrigger [id=1_com.si.totp.app.ClockEvents#increment, interval=30000]: java.util.concurrent.CompletionException: java.lang.IllegalStateException: Toolkit not initialized
        at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:332)
        at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:347)
        at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:874)
        at java.base/java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:887)
        at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2357)
        at java.base/java.util.concurrent.CompletableFuture$MinimalStage.whenComplete(CompletableFuture.java:2948)
        at io.quarkus.scheduler.common.runtime.DefaultInvoker.invoke(DefaultInvoker.java:24)
        at io.quarkus.scheduler.common.runtime.StatusEmitterInvoker.invoke(StatusEmitterInvoker.java:35)
        at io.quarkus.scheduler.runtime.SimpleScheduler$ScheduledTask.doInvoke(SimpleScheduler.java:416)
        at io.quarkus.scheduler.runtime.SimpleScheduler$ScheduledTask$1.handle(SimpleScheduler.java:397)
        at io.quarkus.scheduler.runtime.SimpleScheduler$ScheduledTask$1.handle(SimpleScheduler.java:393)
        at io.vertx.core.impl.ContextBase.lambda$executeBlocking$1(ContextBase.java:180)
        at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:277)
        at io.vertx.core.impl.ContextBase.lambda$internalExecuteBlocking$2(ContextBase.java:199)
        at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:581)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.IllegalStateException: Toolkit not initialized
        at com.sun.javafx.application.PlatformImpl.runLater(PlatformImpl.java:436)
        at com.sun.javafx.application.PlatformImpl.runLater(PlatformImpl.java:431)
        at javafx.application.Platform.runLater(Platform.java:168)
        at com.si.totp.app.ClockEvents.increment(ClockEvents.java:27)
        at com.si.totp.app.ClockEvents_ClientProxy.increment(Unknown Source)
        at com.si.totp.app.ClockEvents_ScheduledInvoker_increment_1c777d19e98d20799c6b72c29b26289f3f771250.invokeBean(Unknown Source)
        ... 15 more
```

So my [quarkus-fx](https://github.com/starksm64/quarkus-fx.git) fork has a forced example of the problem in the samples/clock subdirectory.

What happens is that there is a race condition between scheduler thread and the javafx launcher thread. See the points labelled with // (1), // (2), and // (3):

```
2024-02-06 17:12:18,141 INFO  [io.quarkus] (main) quarkus-fx-sample-using-scheduler 999-SNAPSHOT on JVM (powered by Quarkus 3.2.10.Final) started in 0.333s. 
2024-02-06 17:12:18,141 INFO  [io.quarkus] (main) Profile prod activated. 
2024-02-06 17:12:18,145 INFO  [io.quarkus] (main) Installed features: [cdi, quarkus-fx, scheduler, smallrye-context-propagation, vertx]
java.lang.Exception: Stack trace
        at java.base/java.lang.Thread.dumpStack(Thread.java:2209)
        at io.quarkiverse.fx.QuarkusFxApplication.run(QuarkusFxApplication.java:10) // (1)
        at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:132)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:71)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:44)
        at io.quarkus.runner.GeneratedMain.main(Unknown Source)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:61)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:32)

2024-02-06 17:12:19,084 ERROR [io.qua.sch.com.run.StatusEmitterInvoker] (executor-thread-1) Error occurred while executing task for trigger IntervalTrigger [id=1_io.quarkiverse.fx.sample.ClockEvents#timeIncrement, interval=1000]: java.util.concurrent.CompletionException: java.lang.IllegalStateException: Toolkit not initialized  // (2)
        at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:332)
        at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:347)
        at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:874)
        at java.base/java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:887)
        at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2357)
        at java.base/java.util.concurrent.CompletableFuture$MinimalStage.whenComplete(CompletableFuture.java:2948)
        at io.quarkus.scheduler.common.runtime.DefaultInvoker.invoke(DefaultInvoker.java:24)
        at io.quarkus.scheduler.common.runtime.StatusEmitterInvoker.invoke(StatusEmitterInvoker.java:35)
        at io.quarkus.scheduler.runtime.SimpleScheduler$ScheduledTask.doInvoke(SimpleScheduler.java:416)
        at io.quarkus.scheduler.runtime.SimpleScheduler$ScheduledTask$1.handle(SimpleScheduler.java:397)
        at io.quarkus.scheduler.runtime.SimpleScheduler$ScheduledTask$1.handle(SimpleScheduler.java:393)
        at io.vertx.core.impl.ContextBase.lambda$executeBlocking$1(ContextBase.java:180)
        at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:277)
        at io.vertx.core.impl.ContextBase.lambda$internalExecuteBlocking$2(ContextBase.java:199)
        at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:581)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.IllegalStateException: Toolkit not initialized
        at com.sun.javafx.application.PlatformImpl.runLater(PlatformImpl.java:436)
        at com.sun.javafx.application.PlatformImpl.runLater(PlatformImpl.java:431)
        at javafx.application.Platform.runLater(Platform.java:168)
        at io.quarkiverse.fx.RunOnFxThreadInterceptor.runOnFxThread(RunOnFxThreadInterceptor.java:23)
        at io.quarkiverse.fx.RunOnFxThreadInterceptor_Bean.intercept(Unknown Source)
        at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:42)
        at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:30)
        at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:27)
        at io.quarkiverse.fx.sample.ClockEvents_Subclass.timeIncrement(Unknown Source)
        at io.quarkiverse.fx.sample.ClockEvents_ClientProxy.timeIncrement(Unknown Source)
        at io.quarkiverse.fx.sample.ClockEvents_ScheduledInvoker_timeIncrement_37506911735539830375cffc7b6444bd80624f84.invokeBean(Unknown Source)
        ... 15 more

2024-02-06 17:12:19,155 WARNING [javafx] (JavaFX-Launcher) Unsupported JavaFX configuration: classes were loaded from 'unnamed module @79ca92b9'
2024-02-06 17:12:19,272 INFO  [io.qua.fx.sam.QuarkusFxApp] (JavaFX Application Thread) Begin start: java.lang.Exception
        at io.quarkiverse.fx.sample.QuarkusFxApp.start(QuarkusFxApp.java:26). // (3)
        at io.quarkiverse.fx.sample.QuarkusFxApp_Observer_start_659724dec8591b6562ac2854144bab04e342b063.notify(Unknown Source)
        at io.quarkus.arc.impl.EventImpl$Notifier.notifyObservers(EventImpl.java:346)
        at io.quarkus.arc.impl.EventImpl$Notifier.notify(EventImpl.java:328)
        at io.quarkus.arc.impl.EventImpl.fire(EventImpl.java:82)
        at io.quarkiverse.fx.FxApplication.start(FxApplication.java:19)
        at com.sun.javafx.application.LauncherImpl.lambda$launchApplication1$9(LauncherImpl.java:839)
        at com.sun.javafx.application.PlatformImpl.lambda$runAndWait$12(PlatformImpl.java:483)
        at com.sun.javafx.application.PlatformImpl.lambda$runLater$10(PlatformImpl.java:456)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:400)
        at com.sun.javafx.application.PlatformImpl.lambda$runLater$11(PlatformImpl.java:455)
        at com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:95)

...
```

// (1) this is where the quarkus main entry is being run which calls the javafx Application.launch for the javafx app.

// (2) this is where the scheduler framework has made a call into `@Scheduled` method, which calls Platform.runLater before the javafx app is initialized

// (3) this is where the thread created by the javafx Application.launch calls the io.quarkiverse.fx.FxApplication which intern fires the `@PrimaryStage` event.

The race is between (2) and (3). The scheduler has no knowledge of javafx and its requirements to initialize from (3) before any interaction with Platform.runLater() can happen.

Originally I was directly calling Platform.runLater() from the scheduler method. The use of the `@RunOnFxThread` does allow the possibility of introducing a startup condition to the `RunOnFxThreadInterceptor` that would delay the invocation until after the `io.quarkiverse.fx.FxApplication.start()` method finishes.

--- a/deployment/src/main/java/io/quarkiverse/fx/deployment/QuarkusFxExtensionProcessor.java
+++ b/deployment/src/main/java/io/quarkiverse/fx/deployment/QuarkusFxExtensionProcessor.java
@@ -40,9 +40,24 @@ class QuarkusFxExtensionProcessor {
         return new AdditionalBeanBuildItem(FXMLLoaderProducer.class);
     }
 
+    /**
+     * Register the PrimaryStage qualifier
+     *
+     * @return build item for PrimaryStage
+     */
     @BuildStep
     AdditionalBeanBuildItem primaryStage() {
-        return new AdditionalBeanBuildItem(PrimaryStage.class, FxApplication.class);
+        return new AdditionalBeanBuildItem(PrimaryStage.class);
+    }
+
+    /**
+     * Register the FxApplication so the startup latch producer method is found
+     *
+     * @return build item for FxApplication
+     */
+    @BuildStep
+    AdditionalBeanBuildItem fxApplication() {
+        return new AdditionalBeanBuildItem(FxApplication.class);
     }
 
     @BuildStep

--- a/deployment/src/main/java/io/quarkiverse/fx/deployment/QuarkusFxExtensionProcessor.java
+++ b/deployment/src/main/java/io/quarkiverse/fx/deployment/QuarkusFxExtensionProcessor.java
@@ -10,12 +10,7 @@ import org.jboss.jandex.IndexView;
 import org.jboss.jandex.VoidType;
 import org.jboss.logging.Logger;
 
-import io.quarkiverse.fx.FXMLLoaderProducer;
-import io.quarkiverse.fx.FxApplication;
-import io.quarkiverse.fx.PrimaryStage;
-import io.quarkiverse.fx.QuarkusFxApplication;
-import io.quarkiverse.fx.RunOnFxThread;
-import io.quarkiverse.fx.RunOnFxThreadInterceptor;
+import io.quarkiverse.fx.*;
 import io.quarkus.arc.deployment.AdditionalBeanBuildItem;
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
@@ -56,8 +51,8 @@ class QuarkusFxExtensionProcessor {
      * @return build item for FxApplication
      */
     @BuildStep
-    AdditionalBeanBuildItem fxApplication() {
-        return new AdditionalBeanBuildItem(FxApplication.class);
+    AdditionalBeanBuildItem startupLatch() {
+        return new AdditionalBeanBuildItem(StartupLatch.class);
     }
 
     @BuildStep

--- a/deployment/src/main/java/io/quarkiverse/fx/deployment/QuarkusFxExtensionProcessor.java
+++ b/deployment/src/main/java/io/quarkiverse/fx/deployment/QuarkusFxExtensionProcessor.java
@@ -11,6 +11,7 @@ import org.jboss.jandex.VoidType;
 import org.jboss.logging.Logger;
 
 import io.quarkiverse.fx.FXMLLoaderProducer;
+import io.quarkiverse.fx.FxApplication;
 import io.quarkiverse.fx.PrimaryStage;
 import io.quarkiverse.fx.QuarkusFxApplication;
 import io.quarkiverse.fx.RunOnFxThread;
@@ -41,7 +42,7 @@ class QuarkusFxExtensionProcessor {
 
     @BuildStep
     AdditionalBeanBuildItem primaryStage() {
-        return new AdditionalBeanBuildItem(PrimaryStage.class);
+        return new AdditionalBeanBuildItem(PrimaryStage.class, FxApplication.class);
     }
 
     @BuildStep

--- a/runtime/src/main/java/io/quarkiverse/fx/FxApplication.java
+++ b/runtime/src/main/java/io/quarkiverse/fx/FxApplication.java
@@ -1,39 +1,40 @@
 package io.quarkiverse.fx;
 
-import java.util.concurrent.CountDownLatch;
-
 import jakarta.enterprise.context.ApplicationScoped;
-import jakarta.enterprise.inject.Produces;
+import jakarta.enterprise.context.spi.CreationalContext;
+import jakarta.enterprise.inject.spi.Bean;
+import jakarta.enterprise.inject.spi.BeanManager;
 import jakarta.enterprise.inject.spi.CDI;
 import jakarta.enterprise.util.AnnotationLiteral;
 
 import javafx.application.Application;
 import javafx.stage.Stage;
 
-@ApplicationScoped
+/**
+ * A non-CDI bean that is instantiated by the {@linkplain Application} layer. This
+ * calls into the quarkus-fx application bean using a CDI event with {@linkplain Stage}
+ * with a {@linkplain PrimaryStage} qualifier.
+ *
+ * @see PrimaryStage
+ * @see Stage
+ */
 public class FxApplication extends Application {
-    private static CountDownLatch started = new CountDownLatch(1);
-
-    /**
-     * A producer for the FxApplication.started CountDownLatch
-     *
-     * @return FxApplication.started CountDownLatch
-     */
-    @Produces
-    @PrimaryStage
-    CountDownLatch produceStartedFlag() {
-        return started;
-    }
 
     @Override
     public void start(final Stage primaryStage) {
 
-        CDI.current()
-                .getBeanManager()
-                .getEvent()
-                .select(new AnnotationLiteral<PrimaryStage>() {
-                })
-                .fire(primaryStage);
+        // We need to obtain the StartupLatch singleton
+        BeanManager bm = CDI.current().getBeanManager();
+        Bean<StartupLatch> bean = (Bean<StartupLatch>) bm.getBeans(StartupLatch.class).iterator().next();
+        CreationalContext<StartupLatch> ctx = bm.createCreationalContext(bean);
+        StartupLatch started = (StartupLatch) bm.getReference(bean, StartupLatch.class, ctx);
+
+        // Now broadcast the startup event
+        bm
+            .getEvent()
+            .select(new AnnotationLiteral<PrimaryStage>() {})
+            .fire(primaryStage);
+
         // Mark that the application has finished starting.
         started.countDown();
     }

--- a/runtime/src/main/java/io/quarkiverse/fx/FxApplication.java
+++ b/runtime/src/main/java/io/quarkiverse/fx/FxApplication.java
@@ -2,6 +2,7 @@ package io.quarkiverse.fx;
 
 import java.util.concurrent.CountDownLatch;
 
+import jakarta.enterprise.context.ApplicationScoped;
 import jakarta.enterprise.inject.Produces;
 import jakarta.enterprise.inject.spi.CDI;
 import jakarta.enterprise.util.AnnotationLiteral;
@@ -9,11 +10,13 @@ import jakarta.enterprise.util.AnnotationLiteral;
 import javafx.application.Application;
 import javafx.stage.Stage;
 
+@ApplicationScoped
 public class FxApplication extends Application {
     private static CountDownLatch started = new CountDownLatch(1);
 
     /**
      * A producer for the FxApplication.started CountDownLatch
+     *
      * @return FxApplication.started CountDownLatch
      */
     @Produces
@@ -31,7 +34,7 @@ public class FxApplication extends Application {
                 .select(new AnnotationLiteral<PrimaryStage>() {
                 })
                 .fire(primaryStage);
-        // Mark that the applciation start has finished.
+        // Mark that the application has finished starting.
         started.countDown();
     }
 }

--- a/runtime/src/main/java/io/quarkiverse/fx/FxApplication.java
+++ b/runtime/src/main/java/io/quarkiverse/fx/FxApplication.java
@@ -1,5 +1,8 @@
 package io.quarkiverse.fx;
 
+import java.util.concurrent.CountDownLatch;
+
+import jakarta.enterprise.inject.Produces;
 import jakarta.enterprise.inject.spi.CDI;
 import jakarta.enterprise.util.AnnotationLiteral;
 
@@ -7,6 +10,17 @@ import javafx.application.Application;
 import javafx.stage.Stage;
 
 public class FxApplication extends Application {
+    private static CountDownLatch started = new CountDownLatch(1);
+
+    /**
+     * A producer for the FxApplication.started CountDownLatch
+     * @return FxApplication.started CountDownLatch
+     */
+    @Produces
+    @PrimaryStage
+    CountDownLatch produceStartedFlag() {
+        return started;
+    }
 
     @Override
     public void start(final Stage primaryStage) {
@@ -17,5 +31,7 @@ public class FxApplication extends Application {
                 .select(new AnnotationLiteral<PrimaryStage>() {
                 })
                 .fire(primaryStage);
+        // Mark that the applciation start has finished.
+        started.countDown();
     }
 }

--- a/runtime/src/main/java/io/quarkiverse/fx/PrimaryStage.java
+++ b/runtime/src/main/java/io/quarkiverse/fx/PrimaryStage.java
@@ -6,6 +6,7 @@ import java.lang.annotation.RetentionPolicy;
 import java.lang.annotation.Target;
 
 import jakarta.inject.Qualifier;
+
 import javafx.stage.Stage;
 
 /**

--- a/runtime/src/main/java/io/quarkiverse/fx/PrimaryStage.java
+++ b/runtime/src/main/java/io/quarkiverse/fx/PrimaryStage.java
@@ -6,9 +6,20 @@ import java.lang.annotation.RetentionPolicy;
 import java.lang.annotation.Target;
 
 import jakarta.inject.Qualifier;
+import javafx.stage.Stage;
 
+/**
+ * This is used as a qualifier as for receiving the notification main <code>start</code> entry point for all
+ * JavaFX applications has been called.
+ *
+ * It is also used to qualify the CountDownLatch type used by the {@linkplain FxApplication} uses
+ * as a startup notification barrier.
+ *
+ * @see javafx.application.Application#start(Stage)
+ * @see FxApplication#produceStartedFlag()
+ */
 @Qualifier
-@Target(ElementType.PARAMETER)
+@Target({ ElementType.PARAMETER, ElementType.METHOD, ElementType.FIELD })
 @Retention(RetentionPolicy.RUNTIME)
 public @interface PrimaryStage {
 }

--- a/runtime/src/main/java/io/quarkiverse/fx/RunOnFxThreadInterceptor.java
+++ b/runtime/src/main/java/io/quarkiverse/fx/RunOnFxThreadInterceptor.java
@@ -19,14 +19,14 @@ public class RunOnFxThreadInterceptor {
     // The startup latch from FxApplication
     @PrimaryStage
     @Inject
-    CountDownLatch hasStarted;
+    CountDownLatch startupLatch;
 
     @AroundInvoke
     public Object runOnFxThread(final InvocationContext ctx) throws Exception {
         LOGGER.tracef("intercepted %s on thread %s", ctx.getMethod(), Thread.currentThread());
         // Block any thread until the startup latch has been cleared
         // This will return immediately after the FxApplication#start has completed
-        hasStarted.await();
+        startupLatch.await();
 
         if (Platform.isFxApplicationThread()) {
             return ctx.proceed();

--- a/runtime/src/main/java/io/quarkiverse/fx/RunOnFxThreadInterceptor.java
+++ b/runtime/src/main/java/io/quarkiverse/fx/RunOnFxThreadInterceptor.java
@@ -1,7 +1,5 @@
 package io.quarkiverse.fx;
 
-import java.util.concurrent.CountDownLatch;
-
 import jakarta.inject.Inject;
 import jakarta.interceptor.AroundInvoke;
 import jakarta.interceptor.Interceptor;
@@ -16,10 +14,9 @@ import javafx.application.Platform;
 public class RunOnFxThreadInterceptor {
 
     private static final Logger LOGGER = Logger.getLogger(RunOnFxThreadInterceptor.class);
-    // The startup latch from FxApplication
-    @PrimaryStage
+    // The startup latch signalled by FxApplication
     @Inject
-    CountDownLatch hasStarted;
+    StartupLatch hasStarted;
 
     @AroundInvoke
     public Object runOnFxThread(final InvocationContext ctx) throws Exception {

--- a/runtime/src/main/java/io/quarkiverse/fx/RunOnFxThreadInterceptor.java
+++ b/runtime/src/main/java/io/quarkiverse/fx/RunOnFxThreadInterceptor.java
@@ -19,14 +19,14 @@ public class RunOnFxThreadInterceptor {
     // The startup latch from FxApplication
     @PrimaryStage
     @Inject
-    CountDownLatch startupLatch;
+    CountDownLatch hasStarted;
 
     @AroundInvoke
     public Object runOnFxThread(final InvocationContext ctx) throws Exception {
         LOGGER.tracef("intercepted %s on thread %s", ctx.getMethod(), Thread.currentThread());
         // Block any thread until the startup latch has been cleared
         // This will return immediately after the FxApplication#start has completed
-        startupLatch.await();
+        hasStarted.await();
 
         if (Platform.isFxApplicationThread()) {
             return ctx.proceed();

--- a/runtime/src/main/java/io/quarkiverse/fx/RunOnFxThreadInterceptor.java
+++ b/runtime/src/main/java/io/quarkiverse/fx/RunOnFxThreadInterceptor.java
@@ -1,5 +1,8 @@
 package io.quarkiverse.fx;
 
+import java.util.concurrent.CountDownLatch;
+
+import jakarta.inject.Inject;
 import jakarta.interceptor.AroundInvoke;
 import jakarta.interceptor.Interceptor;
 import jakarta.interceptor.InvocationContext;
@@ -13,10 +16,18 @@ import javafx.application.Platform;
 public class RunOnFxThreadInterceptor {
 
     private static final Logger LOGGER = Logger.getLogger(RunOnFxThreadInterceptor.class);
+    // The startup latch from FxApplication
+    @PrimaryStage
+    @Inject
+    CountDownLatch hasStarted;
 
     @AroundInvoke
     public Object runOnFxThread(final InvocationContext ctx) throws Exception {
         LOGGER.tracef("intercepted %s on thread %s", ctx.getMethod(), Thread.currentThread());
+        // Block any thread until the startup latch has been cleared
+        // This will return immediately after the FxApplication#start has completed
+        hasStarted.await();
+
         if (Platform.isFxApplicationThread()) {
             return ctx.proceed();
         } else {

--- a/runtime/src/main/java/io/quarkiverse/fx/StartupLatch.java
+++ b/runtime/src/main/java/io/quarkiverse/fx/StartupLatch.java
@@ -1,0 +1,21 @@
+package io.quarkiverse.fx;
+
+import java.util.concurrent.CountDownLatch;
+
+import jakarta.inject.Singleton;
+
+/**
+ * A singleton for a startup latch used by {@linkplain FxApplication} and {@linkplain RunOnFxThreadInterceptor}
+ */
+@Singleton
+public class StartupLatch {
+    private final CountDownLatch started = new CountDownLatch(1);
+
+    public void await() throws InterruptedException {
+        started.await();
+    }
+
+    public void countDown() {
+        started.countDown();
+    }
+}

--- a/samples/using-scheduler/pom.xml
+++ b/samples/using-scheduler/pom.xml
@@ -1,5 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
     <modelVersion>4.0.0</modelVersion>
 
     <parent>
@@ -12,10 +13,10 @@
     <name>Quarkus FX - Sample - Using Scheduler</name>
 
     <dependencies>
-    <dependency>
-      <groupId>io.quarkus</groupId>
-      <artifactId>quarkus-scheduler</artifactId>
-    </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-scheduler</artifactId>
+        </dependency>
         <dependency>
             <groupId>io.quarkiverse.fx</groupId>
             <artifactId>quarkus-fx</artifactId>

--- a/samples/using-scheduler/pom.xml
+++ b/samples/using-scheduler/pom.xml
@@ -1,0 +1,54 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>io.quarkiverse.fx</groupId>
+        <artifactId>quarkus-fx-parent</artifactId>
+        <version>999-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>quarkus-fx-sample-using-scheduler</artifactId>
+    <name>Quarkus FX - Sample - Using Scheduler</name>
+
+    <dependencies>
+    <dependency>
+      <groupId>io.quarkus</groupId>
+      <artifactId>quarkus-scheduler</artifactId>
+    </dependency>
+        <dependency>
+            <groupId>io.quarkiverse.fx</groupId>
+            <artifactId>quarkus-fx</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+    </dependencies>
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>io.quarkus</groupId>
+                <artifactId>quarkus-maven-plugin</artifactId>
+                <version>${quarkus.version}</version>
+                <executions>
+                    <execution>
+                        <goals>
+                            <goal>build</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+            <plugin>
+                <artifactId>maven-compiler-plugin</artifactId>
+            </plugin>
+            <plugin>
+                <artifactId>maven-surefire-plugin</artifactId>
+                <version>3.2.5</version>
+                <configuration>
+                    <systemPropertyVariables>
+                        <java.util.logging.manager>org.jboss.logmanager.LogManager</java.util.logging.manager>
+                        <maven.home>${maven.home}</maven.home>
+                    </systemPropertyVariables>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
+</project>

--- a/samples/using-scheduler/src/main/java/io/quarkiverse/fx/sample/AppController.java
+++ b/samples/using-scheduler/src/main/java/io/quarkiverse/fx/sample/AppController.java
@@ -1,0 +1,27 @@
+package io.quarkiverse.fx.sample;
+
+import jakarta.enterprise.context.Dependent;
+import jakarta.enterprise.inject.Instance;
+import jakarta.inject.Inject;
+
+import javafx.application.Platform;
+import javafx.fxml.FXML;
+import javafx.fxml.FXMLLoader;
+import javafx.scene.control.Label;
+
+@Dependent
+public class AppController {
+
+    @Inject
+    Instance<FXMLLoader> fxmlLoaderInstance;
+    @FXML
+    private Label timeLabel;
+    @FXML
+    private void initialize() {
+        //
+    }
+
+    public void onMessage(String timeString) {
+        Platform.runLater(() -> timeLabel.setText(timeString));
+    }
+}

--- a/samples/using-scheduler/src/main/java/io/quarkiverse/fx/sample/AppController.java
+++ b/samples/using-scheduler/src/main/java/io/quarkiverse/fx/sample/AppController.java
@@ -16,6 +16,7 @@ public class AppController {
     Instance<FXMLLoader> fxmlLoaderInstance;
     @FXML
     private Label timeLabel;
+
     @FXML
     private void initialize() {
         //

--- a/samples/using-scheduler/src/main/java/io/quarkiverse/fx/sample/ClockEvents.java
+++ b/samples/using-scheduler/src/main/java/io/quarkiverse/fx/sample/ClockEvents.java
@@ -1,0 +1,49 @@
+package io.quarkiverse.fx.sample;
+
+import java.io.IOException;
+
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.enterprise.event.Event;
+import jakarta.inject.Inject;
+
+import io.quarkiverse.fx.RunOnFxThread;
+import io.quarkus.logging.Log;
+import io.quarkus.scheduler.Scheduled;
+import javafx.stage.Stage;
+
+/**
+ * A sample CDI bean that receives Quarkus scheduler events that are routed back to the javafx interface
+ */
+@ApplicationScoped
+public class ClockEvents {
+    @Inject
+    UnixTimeProvider timeProvider;
+    @Inject
+    Event<TimeEvent> timeEvent;
+
+    /**
+     * A scheduler callback method that delegates to method that is to be run on the
+     * JavaFX creates an application thread. This allows for better startup behavior
+     * than direct use of javafx.application.Platform#runLater(...) as the interceptor
+     * associated with @RunOnFxThread is able to synchronize with the
+     * {@linkplain io.quarkiverse.fx.FxApplication#start(javafx.stage.Stage)} completion.
+     */
+    @Scheduled(every = "1s")
+    @RunOnFxThread
+    void timeIncrement() {
+        Log.info("timeIncrement");
+        sendTimeEvent();
+    }
+
+    private void sendTimeEvent() {
+        try {
+            long unixTime = timeProvider.getTime();
+            String timeString = String.format("%016x", unixTime);
+            TimeEvent event = new TimeEvent(unixTime, timeString);
+            Log.infof("%d/%s", unixTime, timeString);
+            timeEvent.fireAsync(event);
+        } catch (IOException e) {
+            Log.error("Failed to send TimeEvent", e);
+        }
+    }
+}

--- a/samples/using-scheduler/src/main/java/io/quarkiverse/fx/sample/ClockEvents.java
+++ b/samples/using-scheduler/src/main/java/io/quarkiverse/fx/sample/ClockEvents.java
@@ -41,7 +41,7 @@ public class ClockEvents {
             String timeString = String.format("%016x", unixTime);
             TimeEvent event = new TimeEvent(unixTime, timeString);
             Log.infof("%d/%s", unixTime, timeString);
-            timeEvent.fireAsync(event);
+            timeEvent.fire(event);
         } catch (IOException e) {
             Log.error("Failed to send TimeEvent", e);
         }

--- a/samples/using-scheduler/src/main/java/io/quarkiverse/fx/sample/QuarkusFxApp.java
+++ b/samples/using-scheduler/src/main/java/io/quarkiverse/fx/sample/QuarkusFxApp.java
@@ -1,0 +1,51 @@
+package io.quarkiverse.fx.sample;
+
+import java.io.InputStream;
+
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.enterprise.event.Observes;
+import jakarta.enterprise.event.ObservesAsync;
+import jakarta.inject.Inject;
+
+import io.quarkiverse.fx.PrimaryStage;
+import io.quarkus.logging.Log;
+import javafx.application.Platform;
+import javafx.fxml.FXMLLoader;
+import javafx.scene.Parent;
+import javafx.scene.Scene;
+import javafx.stage.Stage;
+
+@ApplicationScoped
+public class QuarkusFxApp {
+
+    @Inject
+    FXMLLoader fxmlLoader;
+    private AppController appController;
+
+    public void start(@Observes @PrimaryStage final Stage stage) {
+        Log.info("Begin start");
+        stage.setOnCloseRequest(event -> {
+            Platform.exit();
+            System.exit(0);
+        });
+
+        try {
+            InputStream fxml = this.getClass().getResourceAsStream("/app.fxml");
+            Parent fxmlParent = this.fxmlLoader.load(fxml);
+            appController = this.fxmlLoader.getController();
+
+            Scene scene = new Scene(fxmlParent, 300, 200);
+            stage.setScene(scene);
+            stage.setTitle("Hello Quarkus + JavaFX ! ⏰");
+            stage.show();
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+        Log.info("End start");
+    }
+
+    void onMessage(@ObservesAsync TimeEvent timeEvent) {
+        Log.infof("Time: %s;%s", timeEvent.unixTime(), timeEvent.timeString());
+        appController.onMessage(timeEvent.timeString());
+    }
+}

--- a/samples/using-scheduler/src/main/java/io/quarkiverse/fx/sample/QuarkusFxApp.java
+++ b/samples/using-scheduler/src/main/java/io/quarkiverse/fx/sample/QuarkusFxApp.java
@@ -4,7 +4,6 @@ import java.io.InputStream;
 
 import jakarta.enterprise.context.ApplicationScoped;
 import jakarta.enterprise.event.Observes;
-import jakarta.enterprise.event.ObservesAsync;
 import jakarta.inject.Inject;
 
 import io.quarkiverse.fx.PrimaryStage;
@@ -44,7 +43,7 @@ public class QuarkusFxApp {
         Log.info("End start");
     }
 
-    void onMessage(@ObservesAsync TimeEvent timeEvent) {
+    void onMessage(@Observes TimeEvent timeEvent) {
         Log.infof("Time: %s;%s", timeEvent.unixTime(), timeEvent.timeString());
         appController.onMessage(timeEvent.timeString());
     }

--- a/samples/using-scheduler/src/main/java/io/quarkiverse/fx/sample/SystemUnixTimeProvider.java
+++ b/samples/using-scheduler/src/main/java/io/quarkiverse/fx/sample/SystemUnixTimeProvider.java
@@ -1,0 +1,13 @@
+package io.quarkiverse.fx.sample;
+
+import java.time.Instant;
+
+import jakarta.enterprise.context.ApplicationScoped;
+
+@ApplicationScoped
+public class SystemUnixTimeProvider implements UnixTimeProvider {
+    @Override
+    public long getTime() {
+        return Instant.now().getEpochSecond();
+    }
+}

--- a/samples/using-scheduler/src/main/java/io/quarkiverse/fx/sample/TimeEvent.java
+++ b/samples/using-scheduler/src/main/java/io/quarkiverse/fx/sample/TimeEvent.java
@@ -1,0 +1,4 @@
+package io.quarkiverse.fx.sample;
+
+public record TimeEvent(long unixTime, String timeString) {
+}

--- a/samples/using-scheduler/src/main/java/io/quarkiverse/fx/sample/UnixTimeProvider.java
+++ b/samples/using-scheduler/src/main/java/io/quarkiverse/fx/sample/UnixTimeProvider.java
@@ -1,0 +1,8 @@
+package io.quarkiverse.fx.sample;
+
+import java.io.IOException;
+
+public interface UnixTimeProvider {
+    // Provides the seconds elapsed since the Unix epoch 1970-01-01T00:00:00Z
+    long getTime() throws IOException;
+}

--- a/samples/using-scheduler/src/main/resources/app.fxml
+++ b/samples/using-scheduler/src/main/resources/app.fxml
@@ -1,0 +1,53 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<!--
+  Copyright (c) 2015, 2019, Gluon and/or its affiliates.
+  All rights reserved. Use is subject to license terms.
+
+  This file is available and licensed under the following license:
+
+  Redistribution and use in source and binary forms, with or without
+  modification, are permitted provided that the following conditions
+  are met:
+
+  - Redistributions of source code must retain the above copyright
+    notice, this list of conditions and the following disclaimer.
+  - Redistributions in binary form must reproduce the above copyright
+    notice, this list of conditions and the following disclaimer in
+    the documentation and/or other materials provided with the distribution.
+  - Neither the name of Oracle Corporation nor the names of its
+    contributors may be used to endorse or promote products derived
+    from this software without specific prior written permission.
+
+  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+  OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+-->
+
+<?import javafx.scene.control.Label?>
+<?import javafx.scene.control.TitledPane?>
+<?import javafx.scene.layout.BorderPane?>
+<?import javafx.scene.text.Font?>
+
+
+<TitledPane text="HexClock" xmlns="http://javafx.com/javafx/21" xmlns:fx="http://javafx.com/fxml/1" fx:controller="io.quarkiverse.fx.sample.AppController">
+   <content>
+      <BorderPane maxHeight="-Infinity" maxWidth="-Infinity" minHeight="-Infinity" minWidth="-Infinity" prefHeight="200.0" prefWidth="300.0">
+         <center>
+            <Label fx:id="timeLabel" text="0123456789" BorderPane.alignment="CENTER">
+               <font>
+                  <Font size="24.0" />
+               </font>
+            </Label>
+         </center>
+      </BorderPane>
+   </content>
+</TitledPane>

--- a/samples/using-scheduler/src/main/resources/application.properties
+++ b/samples/using-scheduler/src/main/resources/application.properties
@@ -1,0 +1,1 @@
+quarkus.log.level=DEBUG

--- a/samples/using-scheduler/src/main/resources/application.properties
+++ b/samples/using-scheduler/src/main/resources/application.properties
@@ -1,1 +1,2 @@
-quarkus.log.level=DEBUG
+# To enable more information about the application startup
+#quarkus.log.level=DEBUG

